File: load-fonts_webgl10-typescript/src/client/Engine/Resources/DefaultResources.ts
import ResourceMap from "./ResourceMap";
import TextFileLoader, { FileType } from "./TextFileLoader";
import ColorShaderProgram from "../ShaderPrograms/ColorShaderProgram";
import TextureShaderProgram from "../ShaderPrograms/TextureShaderProgram";
import SpriteShaderProgram from "../ShaderPrograms/SpriteShaderProgram";
import Fonts from "./Fonts";

export default class DefaultResources
{
    private static _instance: DefaultResources = null;
    private readonly _defaultFontName = "assets/fonts/system-default-font";
    private readonly _vColorShaderName = "assets/shaders/vColor.glsl";
    private readonly _fColorShaderName = "assets/shaders/fColor.glsl";
    private readonly _vTextureShaderName = "assets/shaders/vTexture.glsl";
    private readonly _fTextureShaderName = "assets/shaders/fTexture.glsl";
    private _colorShaderProgram: ColorShaderProgram;
    private _textureShaderProgram: TextureShaderProgram;
    private _spriteShaderProgram: SpriteShaderProgram;

    private constructor() { }

    public static get Instance(): DefaultResources
    {
        if (this._instance === null)
        {
            this._instance = new DefaultResources();
        }
        return this._instance;
    }

    public Initialize(callbackFunction: () => void): void
    {
        TextFileLoader.Instance.LoadTextFile(this._vColorShaderName, FileType.TextFile, null);
        TextFileLoader.Instance.LoadTextFile(this._fColorShaderName, FileType.TextFile, null);

        TextFileLoader.Instance.LoadTextFile(this._vTextureShaderName, FileType.TextFile, null);
        TextFileLoader.Instance.LoadTextFile(this._fTextureShaderName, FileType.TextFile, null);

        Fonts.Instance.LoadFont(this._defaultFontName);

        ResourceMap.Instance.SetLoadCompleteCallback(() => { this.CreateShaderPrograms(callbackFunction); });
    }

    public get ColorShaderProgram(): ColorShaderProgram
    {
        return this._colorShaderProgram;
    }

    public get TextureShaderProgram(): TextureShaderProgram
    {
        return this._textureShaderProgram;
    }

    public get SpriteShaderProgram(): SpriteShaderProgram
    {
        return this._spriteShaderProgram;
    }

    public get DefaultFontName(): string
    {
        return this._defaultFontName;
    }

    private CreateShaderPrograms(callbackFunction: () => void): void
    {
        ResourceMap.Instance.SetLoadCompleteCallback(null);

        this._colorShaderProgram = new ColorShaderProgram(this._vColorShaderName, this._fColorShaderName);
        this._textureShaderProgram = new TextureShaderProgram(this._vTextureShaderName, this._fTextureShaderName);
        this._spriteShaderProgram = new SpriteShaderProgram(this._vTextureShaderName, this._fTextureShaderName);
        callbackFunction();
    }
}

File: LoadFonts_OpenTkOpenGL31CSharp/LoadFonts/Engine/Core/Resources/DefaultResources.cs

using OpenTK.Graphics;
using System;

namespace Engine
{
    class DefaultResources
    {
        private static DefaultResources _instance;
        private readonly string _defaultFontName = "Assets/Fonts/system-default-font";
        private readonly string _vColorShaderName = "Assets/Shaders/vColor.glsl";
        private readonly string _fColorShaderName = "Assets/Shaders/fColor.glsl";
        private readonly string _vTextureShaderName = "Assets/Shaders/vTexture.glsl";
        private readonly string _fTextureShaderName = "Assets/Shaders/fTexture.glsl";

        private ColorShaderProgram _colorShaderProgram;
        private TextureShaderProgram _textureShaderProgram;
        private SpriteShaderProgram _spriteShaderProgram;

        private DefaultResources() { }

        public static DefaultResources Instance
        {
            get
            {
                if (_instance == null)
                {
                    _instance = new DefaultResources();
                }
                return _instance;
            }
        }

        public void Initialize(Action callback)
        {
            TextFileLoader.Instance.LoadTextFile(_vColorShaderName, FileType.TextFile, null);
            TextFileLoader.Instance.LoadTextFile(_fColorShaderName, FileType.TextFile, null);

            TextFileLoader.Instance.LoadTextFile(_vTextureShaderName, FileType.TextFile, null);
            TextFileLoader.Instance.LoadTextFile(_fTextureShaderName, FileType.TextFile, null);

            Fonts.Instance.LoadFont(_defaultFontName);
            
            ResourceMap.Instance.SetLoadCompleteCallback(() => { CreateShaderPrograms(callback); });
        }

        public ColorShaderProgram ColorShaderProgram
        {
            get
            {
                return _colorShaderProgram;
            }
        }

        public TextureShaderProgram TextureShaderProgram
        {
            get
            {
                return _textureShaderProgram;
            }
        }

        public SpriteShaderProgram SpriteShaderProgram
        {
            get
            {
                return _spriteShaderProgram;
            }
        }

        public string DefaultFontName
        {
            get
            {
                return _defaultFontName;
            }
        }

        public void CreateShaderPrograms(Action callback)
        {
            _colorShaderProgram = new ColorShaderProgram(_vColorShaderName, _fColorShaderName);
            _textureShaderProgram = new TextureShaderProgram(_vTextureShaderName, _fTextureShaderName);
            _spriteShaderProgram = new SpriteShaderProgram(_vTextureShaderName, _fTextureShaderName);
            callback();
        }
    }
}
